presto-hudi/src/main/java/com/facebook/presto/hudi/split/HudiPartitionSplitGenerator.java

```diff
@@ -0,0 +1,167 @@
+import org.apache.hudi.common.table.view.HoodieTableFileSystemView;
+import org.apache.hudi.common.util.HoodieTimer;
+
+import java.util.ArrayDeque;
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.hudi.HudiMetadata.toMetastoreContext;
+import static com.facebook.presto.hudi.HudiSessionProperties.getMinimumAssignedSplitWeight;
+import static com.facebook.presto.hudi.HudiSessionProperties.getStandardSplitWeightSize;
+import static com.facebook.presto.hudi.HudiSessionProperties.isSizeBasedSplitWeightsEnabled;
+import static com.facebook.presto.hudi.HudiSplitManager.getHudiPartition;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A runnable to take partition names from a queue of partitions to process,
+ * generate Hudi splits from the partition, and add the splits to the async
+ * ConnectorSplit queue.
+ */
+public class HudiPartitionSplitGenerator
+        implements Runnable
+{
+    private static final Logger LOG = Logger.get(HudiPartitionSplitGenerator.class);
+    private final ExtendedHiveMetastore metastore;
+    private final MetastoreContext metastoreContext;
+    private final HudiTableLayoutHandle layout;
+    private final HudiTableHandle table;
+    private final Path tablePath;
+    private final HoodieTableFileSystemView fsView;
+    private final AsyncQueue<ConnectorSplit> asyncQueue;
+    private final ArrayDeque<String> partitionQueue;
+    private final String latestInstant;
+    private final HudiSplitWeightProvider splitWeightProvider;
+    private boolean isRunning;
+
+    public HudiPartitionSplitGenerator(
+            ConnectorSession session,
+            ExtendedHiveMetastore metastore,
+            HudiTableLayoutHandle layout,
+            HoodieTableFileSystemView fsView,
+            AsyncQueue<ConnectorSplit> asyncQueue,
+            ArrayDeque<String> partitionQueue,
+            String latestInstant)
+    {
+        this.metastore = requireNonNull(metastore, "metastore is null");
+        this.metastoreContext = toMetastoreContext(requireNonNull(session, "session is null"));
+        this.layout = requireNonNull(layout, "layout is null");
+        this.table = layout.getTable();
+        this.tablePath = new Path(table.getPath());
+        this.fsView = requireNonNull(fsView, "fsView is null");
+        this.asyncQueue = requireNonNull(asyncQueue, "asyncQueue is null");
+        this.partitionQueue = requireNonNull(partitionQueue, "partitionQueue is null");
+        this.latestInstant = requireNonNull(latestInstant, "latestInstant is null");
+        this.splitWeightProvider = createSplitWeightProvider(requireNonNull(session, "session is null"));
+        this.isRunning = true;
+    }
+
+    @Override
+    public void run()
+    {
+        HoodieTimer timer = new HoodieTimer().startTimer();
+
+        while (isRunning || !partitionQueue.isEmpty()) {
```

Comment:
Should we control for the number of partitions to load in a batch? Otherwise, the split generation can hog up the resources leaving little resource for split processing in case of table with very high partitions.

Reply:
The number of threads to generate splits (`HudiPartitionSplitGenerator`) is controlled by `hudi.split-generator-parallelism`.  I reduce the default value to 4.  Is that what you're concerning?

Reply:
sounds good